## 3/8 Wrap Up Activity Group 7
<hr>

- **Which datasets do you anticipate using in your final project? Are there columns that can be used to merge these datasets into a single dataframe?**

The datasets we plan to use in our final project are the section 12 alcohol licenses dataset and crime incidents data report from boston.gov. As well as the BARI census block groups mapping from Harvard dataverse and potentially the 911 calls ecometrics. The columns that can be used to merge these datasets are based on location. The crime and alcohol datasets include the address and street name of the business or incident. They also include latitude and longitude coordinates which are likely going to be most effective to use to find their position on the map. Since the geographic data reported by the Boston police can be slightly off we can't rely too heavily on it's accuracy, therefore when merging the data together with a spatial join it's important to note that these areas of city blocks represent overall crime in this area, but not at any exact address or coordinate. Other shared columns are the dates. The licenses and crime incidents both include the date of issue or occurrence which represents another possible area to merge the data although this is not our main focus for now.


- **What is the unit(s) of analysis you will be studying for your final project? In other words, what is the unit that each row in your final dataframe(s) will represent? Is this the same level of analysis that is represented in your raw data? If not, what columns in your raw data could be used to aggregate it to the level you need for your study?**

The unit of analysis for our project are city issued alcohol licenses in Boston, MA and incidents of alcohol related crimes. This is because we are looking to see if a neighborhood block group with a higher number of alcohol licenses is more prone to alcohol related crimes (Operating Under the Influence, Liquor Violation, and potentially violence related crimes based on peer reviewed research) than neighborhood block groups with fewer total licenses. In the final dataframe we would have latitude and longitude cordinates of these specific filtered crimes which are mapped onto specific city block groups in the greater Boston area. The level of analysis is not the same as represented in the raw data since our raw data includes the licenses and incidents on an individual basis. As stated above, we will need to do some type of aggregation (most likely spatial join) of crimes and licenses based on location from the raw data, in order to evaluate our research question.

### Load in and compile two datasets to find missing items for final question

In [37]:
import pandas as pd

In [ ]:
# start with first crime csv
df_crime = pd.read_csv('data/2012-2015-crime.csv.gz')

# append all crime csv's into one dataframe
for year in range(2015, 2022):
    # form standard name convention for each year
    file_name = 'data/' + str(year) + '-crime.csv.gz'
    df_crime.append(pd.read_csv(file_name), ignore_index=True)

# license data is all in one
df_alc = pd.read_csv('data/alcohol.csv.gz')

In [39]:
df_alc

license_num historicallicensenum  status                license_type  \
0      LB-418523                  NaN  Active            Retail Malt Wine   
1      LB-427987                  NaN  Active            Retail Malt Wine   
2      LB-423824                  NaN  Active            Retail Malt Wine   
3      LB-410009                  NaN  Active          CV7 All Alc. Airp.   
4      LB-423834                  NaN  Active  BYOB Bring Your Own Bottle   
...          ...                  ...     ...                         ...   
1429   LB-418841                  NaN  Active          CV7 All Alc. Airp.   
1430   LB-382900                  NaN  Active       Farmer-Winery Pouring   
1431   LB-427915                  NaN  Active            Retail Malt Wine   
1432   LB-409585                  NaN  Active      CV7 All Alc. Restrict.   
1433   LB-345458                  NaN  Active             Retail All Alc.   

          issued     expires                               business_name  \
0     2020-07-30  2020-12-31                            OLKA Corporation   
1     2020-07-30  2020-12-31                       The Daily Market, LLC   
2     2020-05-19  2020-12-31                       Verc Enterprises, LLC   
3     2020-03-04  2020-12-31                        SSP America BOS, LLC   
4     2020-02-20  2020-12-31                       Food For Thought, LLC   
...          ...         ...                                         ...   
1429         NaN         NaN           Delaware North Boston Flight, LLC   
1430         NaN         NaN                         Downeast Cider, LLC   
1431         NaN         NaN                          Allium Market, LLC   
1432         NaN         NaN  The Pearl, Seafood Grill and Raw Bar, Inc.   
1433         NaN         NaN                  Wine Gallery - Seaport LLC   

                           dba_name  \
0                          7-Eleven   
1             The Daily Market, LLC   
2                 Verc Allston Gulf   
3                       Wahlburgers   
4     Food For Thought Test Kitchen   
...                             ...   
1429                  North End Bar   
1430       Downeast Cider House LLC   
1431                  Allium Market   
1432                      The Pearl   
1433      Wine Gallery-Spark Coffee   

                                               comments  \
0                                            NO SINGLES   
1                                                   NaN   
2                                            NO SINGLES   
3                                                   NaN   
4                 <STRONG><FONT size=5></FONT></STRONG>   
...                                                 ...   
1429                                                NaN   
1430  Thursdays 5pm 8pm Fridays 9p Saturdays 11am 8p...   
1431                                                NaN   
1432                                                NaN   
1433                                                NaN   

                                      location_comments  ...  \
0     Single floor convenience store approx 3,540 sq...  ...   
1     The premises includes a portion of the basemen...  ...   
2                                                   NaN  ...   
3      3,033 sq. ft. restaurant located within Termi...  ...   
4     One floor, one room, one bathroom, a divided b...  ...   
...                                                 ...  ...   
1429  Restaurant/full bar concept in Terminal B, Spa...  ...   
1430  On one floor approx. 16,000 SF 1 entrance 4 ex...  ...   
1431  176 S.F. Kiosk located within Boston Design Ce...  ...   
1432  On first floor: kitchen, bar, and dining area ...  ...   
1433  981 Sq. ft. Single floor. One entrance and one...  ...   

                                        applicant             manager  \
0                                OLKA Corporation      Oltion Kadilli   
1                           The Daily Market, LLC   John C. McDonough   
2

In [40]:
df_crime

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP  \
0          I192068249          2647              Other   
1          I192061894          1106   Confidence Games   
2          I192038828          1107              Fraud   
3          I192008877          1107              Fraud   
4          I182090828          1102              Fraud   
...               ...           ...                ...   
53592   I080542626-00          3125    Warrant Arrests   
53593   I080542626-00          1848     Drug Violation   
53594   I080542626-00          1848     Drug Violation   
53595   I030217815-08          3125    Warrant Arrests   
53596   I030217815-08           111           Homicide   

                                  OFFENSE_DESCRIPTION DISTRICT REPORTING_AREA  \
0                           THREATS TO DO BODILY HARM       B2            280   
1                     FRAUD - CREDIT CARD / ATM FRAUD      C11            356   
2                               FRAUD - IMPERSONATION       A1            172   
3                               FRAUD - IMPERSONATION      E18            525   
4                     FRAUD - FALSE PRETENSE / SCHEME       D4            159   
...                                               ...      ...            ...   
53592                                  WARRANT ARREST       A1            111   
53593  DRUGS - POSS CLASS B - INTENT TO MFR DIST DISP       A1            111   
53594  DRUGS - POSS CLASS B - INTENT TO MFR DIST DISP       A1            111   
53595                                  WARRANT ARREST      E18            520   
53596             MURDER, NON-NEGLIGIENT MANSLAUGHTER      E18            520   

      SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH DAY_OF_WEEK  HOUR  \
0          NaN  2015-08-28 10:20:00  2015      8      Friday    10   
1          NaN  2015-08-20 00:00:00  2015      8    Thursday     0   
2          NaN  2015-11-02 12:24:00  2015     11      Monday    12   
3          NaN  2015-07-31 10:00:00  2015      7      Friday    10   
4          NaN  2015-12-01 12:00:00  2015     12     Tuesday    12   
...        ...                  ...   ...    ...         ...   ...   
53592      NaN  2015-08-12 12:00:00  2015      8   Wednesday    12   
53593      NaN  2015-08-12 12:00:00  2015      8   Wednesday    12   
53594      NaN  2015-08-12 12:00:00  2015      8   Wednesday    12   
53595      NaN  2015-07-09 13:38:00  2015      7    Thursday    13   
53596      NaN  2015-07-09 13:38:00  2015      7    Thursday    13   

         UCR_PART         STREET        Lat       Long  \
0        Part Two  WASHINGTON ST  42.330119 -71.084251   
1        Part Two     CHARLES ST  42.300605 -71.061268   
2        Part Two      ALBANY ST  42.334288 -71.072395   
3        Part Two     WINGATE RD  42.237009 -71.129566   
4        Part Two       UPTON ST  42.342432 -71.072258   
...           ...            ...        ...        ...   
53592  Part Three    BOYLSTON ST  42.352312 -71.063705   
53593    Part Two    BOYLSTON ST  42.352312 -71.063705   
53594    Part Two    BOYLSTON ST  42.352312 -71.063705   
53595  Part Three       RIVER ST  42.255926 -71.123172   
53596    Part One       RIVER ST  42.255926 -71.123172   

                          Location  
0      (42.33011862, -71.08425106)  
1      (42.30060543, -71.06126785)  
2      (42.33428841, -71.07239518)  
3      (42.23700950, -71.12956606)  
4      (42.34243222, -71.07225766)  
...                            ...  
53592  (42.35231190, -71.06370510)  
53593  (42.35231190, -71.06370510)  
53594  (42.35231190, -71.06370510)  
53595  (42.25592648, -71.12317207)  
53596  (42.25592648, -71.12317207)  

[53597 rows x 17 columns]

In [41]:
# shows types of criminal offenses, we can use Liquor Violations and Operating Under the Influence
# (or other based on research linking crime and alcohol)
df_crime['OFFENSE_CODE_GROUP'].unique()

array(['Other', 'Confidence Games', 'Fraud', 'Harassment',
       'Property Lost', 'Criminal Harassment', 'Larceny',
       'Auto Theft Recovery', 'Larceny From Motor Vehicle',
       'Investigate Person', 'Police Service Incidents',
       'Warrant Arrests', 'Auto Theft', 'Restraining Order Violations',
       'Property Found', 'Robbery', 'Firearm Violations',
       'Medical Assistance', 'Simple Assault', 'Missing Person Located',
       'Missing Person Reported', 'Embezzlement',
       'Motor Vehicle Accident Response', 'Residential Burglary',
       'Counterfeiting', 'Prisoner Related Incidents', 'Vandalism',
       'Investigate Property', 'License Plate Related Incidents',
       'Aggravated Assault', 'Offenses Against Child / Family',
       'Firearm Discovery', 'Verbal Disputes', 'Search Warrants',
       'Violations', 'Landlord/Tenant Disputes', 'Commercial Burglary',
       'Evading Fare', 'Towed', 'Other Burglary', 'License Violation',
       'Drug Violation', 'Fire Related R

### Find percentage of missing data from key rows

In [42]:
print("Percentage of missing 'Lat' data in crime dataset:")
print(sum(df_crime.Lat.isna()) / len(df_crime)*100,'%')

print("Percentage of missing 'Long' data in crime dataset:")
print(sum(df_crime.Long.isna()) / len(df_crime)*100, '%')

print("Percentage of missing 'OFFENSE_CODE_GROUP' data in crime dataset:")
print(sum(df_crime.OFFENSE_CODE_GROUP.isna()) / len(df_crime)*100, '%')

print("Percentage of missing 'OFFENSE_CODE' data in crime dataset:")
print(sum(df_crime.OFFENSE_CODE.isna()) / len(df_crime)*100, '%')

print("Percentage of missing 'OCCURRED_ON_DATE' data in crime dataset:")
print(sum(df_crime.OCCURRED_ON_DATE.isna()) / len(df_crime)*100, '%')

print('-' * 50)

print("Percentage of missing 'gpsx' data in liquor license dataset:")
print(sum(df_alc.gpsx.isna()) / len(df_crime)*100, '%')

print("Percentage of missing 'gpsy' data in liquor license dataset:")
print(sum(df_alc.gpsy.isna()) / len(df_crime)*100, '%')

print("Percentage of missing 'ISSUED' data in liquor license dataset:")
print(sum(df_alc.issued.isna()) / len(df_crime)*100, '%')

print("Percentage of missing 'license_type' data in liquor license dataset:")
print(sum(df_alc.license_type.isna()) / len(df_crime)*100, '%')

Percentage of missing 'Lat' data in crime dataset:
3.970371476015449 %
Percentage of missing 'Long' data in crime dataset:
3.970371476015449 %
Percentage of missing 'OFFENSE_CODE_GROUP' data in crime dataset:
0.0 %
Percentage of missing 'OFFENSE_CODE' data in crime dataset:
0.0 %
Percentage of missing 'OCCURRED_ON_DATE' data in crime dataset:
0.0 %
--------------------------------------------------
Percentage of missing 'gpsx' data in liquor license dataset:
0.09142302740825047 %
Percentage of missing 'gpsy' data in liquor license dataset:
0.09142302740825047 %
Percentage of missing 'ISSUED' data in liquor license dataset:
0.013060432486892924 %
Percentage of missing 'license_type' data in liquor license dataset:
0.0 %


- **If you have the data for your project in hand: Evaluate your main columns of interest for missing data. What percentage of rows are missing data? Can you see any patterns in which rows are missing data?**

The main columns of interest chosen in the evaluation above are the location data of both the licenses and crime incidents and various other important factors. The location data for both is split up into latitude and longitude coordinates where some rows have both values NA. In the crime dataset 3.97% of rows contain missing location data and in the licenses dataset 0.09% contained missing location data. Given the crime dataset contains nearly 35x more rows than the licenses, with ~4% of crimes missing a location we are still left with a large number of crime incidents (51,000+) with a location. The other important features for crime including the offense code and occurence date contain no NA values. The issued date column for the licenses contained 0.013% NA values which, similar to the location, is a very small portion of the data. The licenses type column contained 0% NA values.

A general pattern observed in the missing data is that latitude and longitude columns are missing the same data, which is helpful because it means that either the entire location was recorded or not at all. If it was not recorded we can exclude this point. Another note is that in the future we will need additional techniques beyond just checking NA for the location. We could also check if the location is outside of Boston, since we only want to include crimes in this area, and there could be rows with errors in the latitude and longitude that we don't want included in our analysis, but aren't NA.